### ISSUE TITLE
Bug: `no-loss-of-precision` false positive

### ISSUE THREAD

**Issue by jjspace** [eslint/eslint#15767]
### Environment

Node version: 16.3.1
npm version: 8.1.2
Local ESLint version: 8.12.0
Global ESLint version: none
Operating System: Kubuntu 20

### What parser are you using?

Default (Espree)

### What did you do?

```js
const test = 555.9771118164062
```

[eslint demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50LWRpc2FibGUgbm8tdW51c2VkLXZhcnMgKi9cblxuLy8gYm90aCBvZiB0aGVzZSB2YWx1ZXMgd29yayBqdXN0IGZpbmUgYnV0IGZvciBzb21lIHJlYXNvbiBvbmx5IDJcbi8vIHNob3dzIGFzIGFuIGVycm9yXG5jb25zdCB0ZXN0ID0gNTU1Ljk3NzExMTgxNjQwNjI7XG5jb25zdCB0ZXN0MCA9IDU1NS45NzcxMTE4MTY0MDYwO1xuY29uc3QgdGVzdDIgPSA1NTUuOTc3MTExODE2NDA2MTtcbmNvbnN0IHRlc3QzID0gNTU1Ljk3NzExMTgxNjQwNjM7XG5jb25zdCB0ZXN0NCA9IDU1NS45NzcxMTE4MTY0MDY0O1xuY29uc3QgdGVzdDUgPSA1NTUuOTc3MTExODE2NDA2NTtcbmNvbnN0IHRlc3Q2ID0gNTU1Ljk3NzExMTgxNjQwNjY7XG5jb25zdCB0ZXN0NyA9IDU1NS45NzcxMTE4MTY0MDY3O1xuY29uc3QgdGVzdDggPSA1NTUuOTc3MTExODE2NDA2ODtcbmNvbnN0IHRlc3Q5ID0gNTU1Ljk3NzExMTgxNjQwNjk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLWxvc3Mtb2YtcHJlY2lzaW9uIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNldHRlci1yZXR1cm4iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW5zYWZlLW9wdGlvbmFsLWNoYWluaW5nIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=)

### What did you expect to happen?

This value is actually fine from my testing in chrome, brave, firefox and node console. it shouldn't be marked as an error

### What actually happened?

eslint reported a problem with the `no-loss-of-precision` rule

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @jjspace, thanks for the issue!

> This value is actually fine from my testing in chrome, brave, firefox and node console. it shouldn't be marked as an error

Can you share the tests that show that `555.9771118164062` literal doesn't lose precision?

I'm getting this in chrome v100.0.4896.75 console:

```
> const test = 555.9771118164062
undefined
> test.toPrecision(16)
'555.9771118164063'
> test === 555.9771118164063
true
```

**Comment by ljharb** [comment]
@mdjermanovic is the rule concerned with `.toPrecision`? or is it concerned with the parsing itself - meaning, `String(555.9771118164062) === '555.9771118164062'` indicates no precision loss.

**Comment by mdjermanovic** [comment]
> @mdjermanovic is the rule concerned with `.toPrecision`? or is it concerned with the parsing itself - meaning, `String(555.9771118164062) === '555.9771118164062'` indicates no precision loss.

Basically, it checks `"555.9771118164062" === 555.9771118164062.toPrecision(16)` and reports if that's `false`.

(technically, both sides are normalized and compared as `"5.559771118164062e2" === "5.559771118164063e2"`)

`raw` (literal's text in the source code) should be the same as `value.toPrecision(<number of digits in raw>)`. If it isn't, the rule considers that `raw` loses precision.

**Comment by ljharb** [comment]
hmm - that's a bit confusing to me.

There's some number literals that can't *parse* correctly - for example, `9007199254740993`, which ends up indistinguishable from `9007199254740992`. Those are the kinds I'd want to avoid with eslint - I hardly ever use `.toPrecision`, and i usually rely on `toString` output, so this input is something I'd want to allow.

Given that https://eslint.org/docs/rules/no-loss-of-precision says 
> This rule would disallow the use of number literals that immediately lose precision at runtime when converted to a JS Number due to 64-bit floating-point rounding.

the implication is that its purpose matches my interpretation, and that using `.toPrecision` inside the rule's implementation is just a bug.

**Comment by nzakas** [comment]
The original rule proposal is helpful for this:
https://github.com/eslint/eslint/issues/11279

@ljharb what would you propose instead of toPrecision?

**Comment by ljharb** [comment]
I’m not sure - I’d need to check a bunch of test cases. I’d maybe try: if the toPrecision or the toString is the same, it’s fine, else it’s warned on?

**Comment by mdjermanovic** [comment]
> There's some number literals that can't _parse_ correctly - for example, `9007199254740993`, which ends up indistinguishable from `9007199254740992`. Those are the kinds I'd want to avoid with eslint - I hardly ever use `.toPrecision`, and i usually rely on `toString` output, so this input is something I'd want to allow.

This rule is not intended to check how the number looks when converted to a string. The assumption is that the number will be used in calculations and that the desired precision expressed by writing a certain number of digits is important for those calculations (i.e. all digits in the number literal are significant for those calculations). The rule warns when JS does not support the desired precision (the number value rounded to the desired precision does not match the literal on all digits) which implies that the calculations will not be accurate as expected. Using the builtin method `toPrecision` inside a rule named `no-loss-of-precision` doesn't look like a bug :)

**Comment by jjspace** [comment]
> Can you share the tests that show that `555.9771118164062` literal doesn't lose precision?

Please excuse my naivete in posting this without digging deeper into understanding precision in JS beyond reading the rule description and making assumptions. My "testing" was just to the simple extent of setting the value and making sure it was "preserved" in usage in dev consoles like below

![2022-04-08_11-45](https://user-images.githubusercontent.com/8007967/162477230-a04b2551-bcf9-4662-998f-c2176d21d1e4.png)

To me this showed an example in direct opposition to the part of the rule that states "number literals that immediately lose precision at runtime when converted to a JS Number". I didn't know about the `toPrecision` function or that that's how the rule is handled under the surface. I did a little more testing (see below) and this number definitely has issues with precision at such a small scale. However I don't think it's correct to say it "immediately loses precision" as the value does seem to be preserved until you manipulate it in some way... This issue may just be a matter of updating the documentation to better match the behavior under the hood. Also possibly worth adding a link to some documentation discussing how JS handles precision and/or directly to the [`toPrecision`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toPrecision) MDN page so others who see this error and go to ESLint to check it can become more informed.

```js
let test = 555.9771118164062;
console.log(test); // 555.9771118164062
console.log(test.toPrecision(16)); // 555.9771118164063
console.log(test === 555.9771118164062); // true  <-- this contradicts the "immediately lose..."
console.log(test + 0.0000000000001); // 555.9771118164064
console.log(test + 0.0000000000002); // 555.9771118164065
console.log(test + 0.0000000000003); // 555.9771118164066
console.log(test + 0.0000000000004); // 555.9771118164067
console.log(test + 0.0000000000005); // 555.9771118164067 <-- this is weird right? or is it a rounding issue?
console.log(test + 0.0000000000006); // 555.9771118164068
console.log(test + 0.0000000000007); // 555.9771118164069
console.log(test + 0.0000000000008); // 555.977111816407   
console.log(test + 0.0000000000009); // 555.9771118164072

// and just to prove that there are similar issues with a "valid" number by eslint's standards
test = 555.9771118164063; // <-- this does not trigger the eslint rule
console.log(test); // 555.9771118164062  // <-- but that's not the same value
console.log(test.toPrecision(16)); // 555.9771118164063 <-- but this is?
console.log(test === 555.9771118164063); // true
console.log(test + 0.0000000000001); // 555.9771118164064
console.log(test + 0.0000000000002); // 555.9771118164065
console.log(test + 0.0000000000003); // 555.9771118164066
console.log(test + 0.0000000000004); // 555.9771118164067
console.log(test + 0.0000000000005); // 555.9771118164067
console.log(test + 0.0000000000006); // 555.9771118164068
console.log(test + 0.0000000000007); // 555.9771118164069
console.log(test + 0.0000000000008); // 555.977111816407
console.log(test + 0.0000000000009); // 555.9771118164072
```

**Comment by nzakas** [comment]
> This issue may just be a matter of updating the documentation to better match the behavior under the hood. Also possibly worth adding a link to some documentation discussing how JS handles precision and/or directly to the [toPrecision](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toPrecision) MDN page so others who see this error and go to ESLint to check it can become more informed.

👍

**Comment by ljharb** [comment]
If the decision is to update the docs, then the rule’s usefulness decreases from “any number usage” to “any toPrecision usage”, which is much much rarer in my experience.

**Comment by nzakas** [comment]
@ljharb again, what would you suggest as an alternative? It’s a lot more helpful to give us options than to just say you don’t like something.

**Comment by mdjermanovic** [comment]
> Just to piggyback on this issue, false positives are reported for number literals ending with `0.`, such as:
> 
> ```js
> const x = 10.;
> const y = 1234567890..toFixed(0);
> ```
> 
> Both values are obviously safe precision-wise.

@xemlock this looks like a different problem, could you please open a separate issue?

**Comment by ljharb** [comment]
here's the suggestion: https://github.com/eslint/eslint/issues/15767#issuecomment-1097455674

**Comment by nzakas** [comment]
It looks like we never really came to a conclusion here. The question is: do we update the docs to reflect the current behavior or do we change how the rule works?

Im more inclined to just update the docs as this rule has been around for a while and changing it now might be confusing. 

@eslint/eslint-tsc what do you all think?

**Comment by btmills** [comment]
Reading through the original rule proposal, specifically how it differs from https://github.com/eslint/eslint/issues/8447, it seems to me the intention was always to make this about significant digits of precision, but we didn't use that precise (ha) mathematical term in the docs. I'm in favor of updating the docs to reflect the rule proposal and implementation. I might rephrase to something like:

> This rule would disallow the use of number literals whose significant digits immediately lose precision at runtime when converted to a JS Number due to 64-bit floating-point rounding.

What's been discussed in #8447 and again here sounds like a separate `no-unrepresentable-number` rule.

**Comment by ljharb** [comment]
@btmills that's certainly a path forward. altho "lose precision at runtime" isn't accurate, since `.toPrecision` is not the sole, or primary, arbiter of what precision a number has.

**Comment by jjspace** [comment]
I will point out again what I said in my follow up [comment](https://github.com/eslint/eslint/issues/15767#issuecomment-1093041405) above. The "immediately lose precision at runtime" was the _main_ cause of my confusion and why I thought my original value was a false positive. I don't think it's accurate or helpful to keep that wording in the docs, however they're updated. See my comment above for values I would expect to break the "immediately at runtime" part of the rule.

**Comment by nzakas** [comment]
@jjspace does removing the word “immediately” solve the problem for you?

**Comment by jjspace** [comment]
@nzakas Sorry for forgetting to following up on this. I think removing "immediately" would help. Then I wouldn't be as inclined to think the behavior of https://github.com/eslint/eslint/issues/15767#issuecomment-1093041405 was wrong, just that this value was at some point losing precision? I think it would also help to include the "significant digits" part that @btmills [suggested](https://github.com/eslint/eslint/issues/15767#issuecomment-1179787607). I think it would also be good to include some sort of comment in the docs about how this is checked under the hood with `toPrecision` so future devs can understand it better if @ljharb's comment is accurate that:
> "lose precision at runtime" isn't accurate, since .toPrecision is not the sole, or primary, arbiter of what precision a number has.

I do agree that leaving the rule as is functionally may be better here if people have already been using it to avoid confusion. It honestly has only come up one more time for me since I originally opened this and in that case it was actually accurate that the number it was flagging was changed "immediately" when running the program, even doing my very basic "does this work in the dev console" check.

**Comment by nzakas** [comment]
Okay, then let’s just remove “immediately” from the docs.
